resources/templates/md/posts/2016-05-02-multimethods-protocols-explained.md

```diff
@@ -0,0 +1,486 @@
+want to know how the kitchen actually will look like. You only know that it
+will occupy a given space inside the house. In this sense a _kitchen_ is an 
+abstraction. The same goes for objects inside a room: you may know that you will
+need a bed, a wordrobe and a desk and where they will be placed but at this point
+you don't need to know what concrete bed or desk you are going to buy. 
+All these objects act as abstractions.
+
+Another good analogy is a trip from point A to B: imagine you are planning a
+trip and know that upon arrival to the airport you will rent a car and drive to
+the city center. That's all you know: that you will rent **a** car and you will
+_drive_ it. You will decide at "run-time" which specific car it will be but
+during planning time you don't need to worry about it.
+
+This approach has two advantages:
+
+1) You don't need to worry about the details too soon, you just need to have the
+big picture.
+
+2) You can change or extend your _concrete_ choices: say you bought a metal desk
+but for aesthetic reasons you want to change it to a wooden one. And of course
+you can, because both are just desks and thanks to the good design you did initially
+almost any desk would fit into the room.
+
+As we know, programming imitates life (or was it the other way around?) and it
+has its tools for creating abstractions, too. Usually we reason about
+algorithms (_behaviors_), implemented by functions or methods, just like the
+_drive_ example above.
+
+In Java the main abstraction tool is the `interface`, which contains a set
+of methods without body which other classes can implement. The advantage of
+course is that then we can use the interface instead of a concrete type in our code
+(see point 1 above) and then create many classes (see point 2) which
+provide different implementation of the same operation.
+
+Abstractions are all about **reusability** and **extensibility**, so let's see
+what tools Clojure provides to deal with them.
+
+## Multimethods
+
+A multimethod is nothing more than a way to group together functions under one
+name and select which one to execute based on another function, called
+_dispatching function_. This could be not very clear so let's see an example.
+
+```clojure
+(defmulti tribute (fn [musician] (:role musician)))
+
+(defmethod tribute :drums
+  [musician]
+  (str (:name musician) " was really great at drums! Just like a metronome!"))
+
+(defmethod tribute :vocal
+  [musician]
+  (str (:name musician) " had such a perfect voice!"))
+
+(tribute {:name "Jim Morrison", :role :vocal})
+;; "Jim Morrison had such a perfect voice!"
+
+(tribute {:name "Keith Moon", :role :drums})
+;; "Keith Moon was really great at drums! Just like a metronome!"
+```
+
+First we create a multimethod `tribute` by using `defmulti`. The _dispatching function_ 
+tells us how we want to select the correct implementation, in this
+case we dispatch on `(:role musician)`. This value will be used to determine the
```

Comment:
On the value returned by the expression to be clearer (on feels there is a wizard behind the door)